universal-application-tool-0.0.1/app/services/applicant/predicate/JsonPathPredicateGenerator.java

```diff
@@ -1,20 +1,90 @@
 package services.applicant.predicate;
 
+import static com.google.common.collect.ImmutableMap.toImmutableMap;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import java.util.Optional;
 import services.Path;
+import services.applicant.ApplicantData;
+import services.applicant.RepeatedEntity;
+import services.applicant.exception.InvalidPredicateException;
+import services.applicant.question.ApplicantQuestion;
 import services.program.predicate.LeafOperationExpressionNode;
+import services.question.types.QuestionDefinition;
 
+/** Generates {@link JsonPathPredicate}s based on the current applicant filling out the program. */
 public class JsonPathPredicateGenerator {
 
+  private final ApplicantData applicantData;
+  private final ImmutableMap<Long, QuestionDefinition> questionsById;
+  private final Optional<RepeatedEntity> currentRepeatedContext;
+
+  /**
+   * This cannot be built from a set of {@link ApplicantQuestion}s because the question IDs for
+   * repeated questions are used multiple times. For example, if there is a repeated name question,
+   * that name question is repeated for each entity. Instead, we use {@link QuestionDefinition}s and
+   * pass in the current block's repeated entity context, so we can determine which repeated entity
+   * we are currently on, as well as find the target question used in the predicate.
+   */
+  public JsonPathPredicateGenerator(
+      ApplicantData applicantData,
+      ImmutableList<QuestionDefinition> programQuestions,
+      Optional<RepeatedEntity> currentRepeatedContext) {
+    this.applicantData = applicantData;
+    this.questionsById =
+        programQuestions.stream().collect(toImmutableMap(QuestionDefinition::getId, q -> q));
+    this.currentRepeatedContext = currentRepeatedContext;
+  }
+
   /**
    * Formats a {@link LeafOperationExpressionNode} in JsonPath format: {@code path[?(expression)]}
    *
    * <p>Example: \$.applicant.address[?(@.zip in ["12345", "56789"])]
    */
-  public JsonPathPredicate fromLeafNode(LeafOperationExpressionNode node, Path questionPath) {
+  public JsonPathPredicate fromLeafNode(LeafOperationExpressionNode node)
+      throws InvalidPredicateException {
+    if (!questionsById.containsKey(node.questionId())) {
+      // This means a predicate was incorrectly configured - we are depending upon a question that
+      // does not appear anywhere in this program.
+      throw new InvalidPredicateException(
+          String.format(
+              "Tried to apply a predicate based on question %d, which is not found in this"
+                  + " program.",
+              node.questionId()));
+    }
+
+    QuestionDefinition targetQuestion = questionsById.get(node.questionId());
+    Optional<RepeatedEntity> predicateContext;
+
+    if (targetQuestion.getEnumeratorId().isEmpty()) {
```

Comment:
(maybe blocking): I don't think we need to treat this as a special case. Using `predicateContext.parent()` until it is an empty optional would eventually match `targetQuestion.getEnumeratorId()` if it is an empty optional. 

Getting rid of this as a special case might make the code simpler/easier to maintain. But if this is more readable, then please stick with this!

Reply:
Good callout - I like early exits, which is why I originally put it in. Now, I think based on your other comment that this is necessary to differentiate between the "this is a top-level question with an empty context" case and the "this is an incorrectly configured predicate that led to an empty context" case. Thoughts?